neurodamus/target_manager.py

```diff
@@ -269,9 +220,100 @@ def pathways_overlap(self, conn1, conn2, equal_only=False):
+            metype = self._cell_manager.getMEType(gid)
+            result_list.append(metype)
+
+        return result_list
+
+    def gid_to_sections(self, gid):
+        """
+        For a given gid, return a list of section references stored for random access.
+        If the list does not exist, it is built and stored for future use.
+
+        :param gid: GID of the cell
+        :return: SerializedSections object with SectionRefs to every section in the cell
+        """
+        if gid not in self._section_access:
+            cell_ref = self._cell_manager.get_cellref(gid)
+            if cell_ref is None:
+                logging.warning("No cell found for GID: %d", gid)
+                return None
+            result_serial = SerializedSections(cell_ref)
+            self._section_access[gid] = result_serial
+        else:
+            result_serial = self._section_access[gid]
+
+        return result_serial
+
+    def location_to_point(self, gid, isec, ipt, offset):
+        """
+        Given a location for a cell, section id, segment id, and offset into the segment,
+        create a list containing a section reference to there.
+
+        :param gid: GID of the cell
+        :param isec: Section index
+        :param ipt: Distance to start of segment
+        :param offset: Offset distance beyond the ipt (microns)
+        :return: List with 1 item, where the synapse should go
+        """
+        # Soma connection, just zero it
+        if offset < 0:
+            offset = 0
+
+        result_point = Nd.TPointList(gid)
+        cell_sections = self.gid_to_sections(gid)
+        if not cell_sections:
+            raise Exception("Getting locations for non-bg sims is not implemented yet...")
+
+        if isec >= cell_sections.num_sections:
+            raise Exception(f"Error: section {isec} out of bounds ({cell_sections.num_sections} "
+                            "total). Morphology section count is low, is this a good morphology?")
+
+        distance = 0.5
+        tmp_section = cell_sections.isec2sec[int(isec)]
+
+        if tmp_section is None:  # Assume we are in LoadBalance mode
+            result_point.append(None, -1)
+        elif ipt == -1:
+            # SYN2/Sonata spec have a pre-calculated distance field.
+            # In such cases, segment (ipt) is -1 and offset is that distance.
+            offset = max(min(offset, 0.9999999), 0.0000001)
+            result_point.append(tmp_section, offset)
+        else:
+            # Adjust for section orientation and calculate distance
+            section = tmp_section.sec
+            if section.orientation() == 1:
+                ipt = section.n3d() - 1 - ipt
+                offset = -offset
+
+            if ipt < section.n3d():
+                distance = (section.arc3d(int(ipt)) + offset) / section.L
+                if distance == 0.0:
+                    distance = 0.0000001
+                if distance >= 1.0:
+                    distance = 0.9999999
```

Comment:
I am wondering if we can use the `max(min(...)...)` as you did above for `offset`. But is it likely that distance < 0 after calculation?